### PR TITLE
[All] Span GestureRecognizers

### LINKED ISSUE #1701: [Enhancement] Allow for commandable span regions
## Rationale

Embedding hyperlinks into a label is useful for many many reasons but the simplest to imagine is an article which contains hyperlinks in the text. It is impractical to expect the text be split up into multiple labels.

## API

This will depend on the Bindable Span spec being implemented https://github.com/xamarin/Xamarin.Forms/issues/1340.

```csharp
public class Span : Base {
  //<Snip/>
  
  public static readonly BindableProperty CommandProperty;
  
  public ICommand Command { get; set; } // BindableProperty backed
}
```

## Expected Result
When a command is set and its CanExecute method evaluates to true, the span with a Command should become tappable and execute the command when tapped.

### Android
Without having looked in any real depth it looks liek this can be achieved with a combination of attributed text and LinkMovementMethod.

### iOS
A TapGestureRecognizer will need to be used in conjunction with checking which part of the label it actually hit. This will be annoying to implement to say the least.

### UWP
https://docs.microsoft.com/en-us/windows/uwp/design/controls-and-patterns/hyperlinks

## Implications for CSS
Not relevant to CSS

## Backward Compatibility

Shouldn't be any issues here.

## Difficulty : Hard

The iOS and Android implementations of this are likely to be quite difficult. Worse we want to make sure that this behavior sits "on top" of any tap gesture recognizer added to the label as a whole. So if you tap a hyperlink, the link should open, if you tap other parts of the label, the gesture recognizer should fire.

### PR DESCRIPTION
**Pull request by adamped**
### Description of Change ###

fixes #1701

This allows Gesture Recognizers to be added to the Span in a Label. This implementation also provides the basis for other controls to also add Gesture Recognizers to their own child elements, such as a potential image map.

Gesture Recognizers were chosen over a command, as this allows the ability for this solution to expand to include things such as long press on text.

```
<Label>
     <Label.FormattedText>
          <FormattedString>
              <Span Text="Span1" />
              <Span Text="Span2 (Tap me)" TextColor="Blue">
                  <Span.GestureRecognizers>
                         <TapGestureRecognizer Command="{Binding TapCommand}" />
                   </Span.GestureRecognizers>
              </Span>
              <Span Text="Span3"/>
          </FormattedString>
     </Label.FormattedText>
</Label>
```

Support for

- Android
- iOS / MacOS
- UWP / WPF

### Bugs Fixed ###

None

### API Changes ###

List all API changes here (or just put None), example:

Added:
 - [Span] GestureRecognizers
 - [IGestureController] GetChildElements
 - [IGestureController] CompositeGestureRecognizers
 - (Internal)[ISpatialElement] Regions

Changed:
 - Span inheritance = Span -> GestureElement -> Element

### Behavioral Changes ###

None

### PR Checklist ###

- [ ] Has tests (if omitted, state reason in description) - Coming soon
- [X] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard
- [ ] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Platform.Android/Renderers/FormattedStringExtensions.cs [hunk 1]
```diff
@@ -55,8 +55,8 @@ public static SpannableString ToAttributed(this FormattedString formattedString,
 #pragma warning disable 618 // We will need to update this when .Font goes away
 					spannable.SetSpan(new FontSpan(span.Font, view), start, end, SpanTypes.InclusiveInclusive);
 #pragma warning restore 618
-				else if (defaultFont != Font.Default)
-					spannable.SetSpan(new FontSpan(defaultFont, view), start, end, SpanTypes.InclusiveInclusive);
+				else
+					spannable.SetSpan(new FontSpan(defaultFont, view), start, end, SpanTypes.InclusiveInclusive);
```
Comment (jerone) [inline Xamarin.Forms.Platform.Android/Renderers/FormattedStringExtensions.cs]: Trailing tabs.

**Comment by samhouts** [comment]
@adamped Would you mind rebasing on master? We've recently removed the docs step that is causing the build to fail. Thanks!

File: Xamarin.Forms.Platform.iOS/EventTracker.cs [hunk 2]
```diff
@@ -133,22 +138,81 @@ protected virtual NativeGestureRecognizer GetNativeRecognizer(IGestureRecognizer
...
+				});
+
+				var uiRecognizer = CreateTapRecognizer(tapRecognizer.NumberOfTapsRequired, returnAction);
+				return uiRecognizer;
+			}
 #endif
 
+			if (recognizer is ChildGestureRecognizer childRecognizer)
+			{
+#if !__MOBILE__
```
Comment (hartez) [inline Xamarin.Forms.Platform.iOS/EventTracker.cs]: Again, not a blocker, but it'd be nice to have this split into separate methods to make it easier to read and maintain.

File: Xamarin.Forms.Platform.WPF/Xamarin.Forms.Platform.WPF.csproj [hunk 3]
```diff
@@ -19,10 +19,10 @@
     <DebugType>full</DebugType>
     <Optimize>false</Optimize>
     <OutputPath>bin\Debug\</OutputPath>
-    <DefineConstants>DEBUG;TRACE</DefineConstants>
+    <DefineConstants>TRACE;DEBUG</DefineConstants>
     <ErrorReport>prompt</ErrorReport>
     <WarningLevel>4</WarningLevel>
-    <LangVersion>6</LangVersion>
+    <LangVersion>7</LangVersion>
```
[the remark this answers is not in the document]
Comment (adamped) [reply Xamarin.Forms.Platform.WPF/Xamarin.Forms.Platform.WPF.csproj]: Not that I could see.

I speculate it was due to the fact that the XF source code used to be restricted to C#6 when WPF was being built.

File: Xamarin.Forms.Core/Label.cs [hunk 4]
```diff
@@ -169,6 +187,83 @@ void OnFormattedTextChanged(object sender, PropertyChangedEventArgs e)
 			OnPropertyChanged("FormattedText");
 		}
 
+		void Span_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
```
[the remark this answers is not in the document]
Comment (adamped) [reply Xamarin.Forms.Core/Label.cs]: Done

File: Xamarin.Forms.Platform.iOS/EventTracker.cs (hunk 2 above)
Comment (adamped) [reply Xamarin.Forms.Platform.iOS/EventTracker.cs]: Done

**Comment by adamped** [comment]
@hartez - made changes as per recommendations

**Comment by ahmedalejo** [comment]
@adamped great work!

**Comment by mnknedim** [comment]
I have run successfully, but the Command method didn't work more than one. So i have research and i found this comment that is  "...You can add a GestureRecogizer to the Label that is displaying the FormattedString but there's no easy way to know where in the string the tap occurred."?

**Comment by PureWeen** [comment]
@mnknedim   a number of changes were made to this. Can you test out 3.2 pre3 and if you are still having issues open up a new issue?

**Comment by mnknedim** [comment]
@PureWeen  Yes, after update of 3.2 pre3  it has run successfully. Thank u so much 👍

**Comment by PureWeen** [comment]
@adrianknight89  if you have more context that'd be helpful. 

David's Little Play things sample uses the binding with 3.3.0 pre1 and it worked fine for me their
https://github.com/davidortinau/TheLittleThingsPlayground

**Pull request merged**